script.js

```diff
@@ -75,26 +75,36 @@ function updateCanvas() {
     drawCharge(charges[i]);
   }
 }
+
 // creates paths array of the form [[[x11, y11], [x12, y12], ...], [[x21, y21], [x22, y22], ...], ...]
 function calculateFieldLines() {
   let paths = [];
   for (let i = 0; i < charges.length; i++) {
-    let startAngle = TAU * Math.random();
-    let angleStep = TAU / Math.floor(1 + Math.abs(charges[i][2]));
-    for (let j = 0; j <= Math.abs(charges[i][2]); j++)
-      paths.push(calculatePath([charges[i][0] + Math.sqrt(Math.abs(charges[i][2])) * Math.cos(startAngle + j * angleStep), charges[i][1] + Math.sqrt(Math.abs(charges[i][2])) * Math.sin(startAngle + j * angleStep), charges[i][2]]));
+    let angle = TAU * Math.random();
+    let radius = Math.sqrt(Math.abs(charges[i][2]));
+    //let angleStep = TAU / Math.floor(1 + Math.abs(charges[i][2]));
+    for (let j = 0; j <= Math.abs(charges[i][2]); j++) {
+      angle +=  TAU / Math.floor(1 + Math.abs(charges[i][2]));
+      if (!alreadyGenerated(paths, charges[i], radius, angle)) paths.push(calculatePath([charges[i][0] + radius * Math.cos(angle), charges[i][1] + radius * Math.sin(angle), charges[i][2]]));
+    }
   }
   return paths;
 }
+// not finished
+function alreadyGenerated(paths, charge, radius, angle) {
+  for (let path of paths)
+  // tests if there any path ends within the charge, and if a path does, if it is close enough for the path at angle to be redundant
+    if ((path.at(-1)[0] - charge[0])**2 + (path.at(-1)[1] - charge[1])**2 < Math.abs(charge[2] - 1) && (path.at(-1)[0] - charge[0] - radius * Math.cos(angle))**2 + (path.at(-1)[1] - charge[1] - radius * Math.sin(angle))**2 < 1 + Math.abs(charge[2])**(-1)) return true;
```

Comment:
This line is a bit of a beast. You need to find some way to break it up. I'd probably try to define some local variables for various parts of the expression with meaningful names. That will make it a lot easier to read. I don't understand enough of what's going on here to guess what those names might be though.